frame/nomination-pools/src/lib.rs

```diff
@@ -0,0 +1,1967 @@
+
+			bonded_pool.try_inc_delegators()?;
+			let points_issued = bonded_pool.try_bond_funds(&who, amount, BondType::Later)?;
+
+			Delegators::insert(
+				who.clone(),
+				Delegator::<T> {
+					pool_id,
+					points: points_issued,
+					// At best the reward pool has the rewards up through the previous era. If the
+					// delegator joins prior to the snapshot they will benefit from the rewards of
+					// the active era despite not contributing to the pool's vote weight. If they
+					// join after the snapshot is taken they will benefit from the rewards of the
+					// next 2 eras because their vote weight will not be counted until the
+					// snapshot in active era + 1.
+					reward_pool_total_earnings: reward_pool.total_earnings,
+					unbonding_era: None,
+				},
+			);
+			bonded_pool.put();
+			Self::deposit_event(Event::<T>::Bonded {
+				delegator: who,
+				pool_id,
+				bonded: amount,
+				joined: true,
+			});
+
+			Ok(())
+		}
+
+		/// Bond `extra` more funds from `origin` into the pool to which they already belong.
+		///
+		/// Additional funds can come from either the free balance of the account, of from the
+		/// accumulated rewards, see [`BondExtra`].
+		// NOTE: this transaction is implemented with the sole purpose of readability and
+		// correctness, not optimization. We read/write several storage items multiple times instead
+		// of just once, in the spirit reusing code.
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn bond_extra(origin: OriginFor<T>, extra: BondExtra<BalanceOf<T>>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let (mut delegator, mut bonded_pool, mut reward_pool) =
+				Self::get_delegator_with_pools(&who)?;
+
+			let (points_issued, bonded) = match extra {
+				BondExtra::FreeBalance(amount) =>
+					(bonded_pool.try_bond_funds(&who, amount, BondType::Later)?, amount),
+				BondExtra::Rewards => {
+					let claimed = Self::do_reward_payout(
+						&who,
+						&mut delegator,
+						&mut bonded_pool,
+						&mut reward_pool,
+					)?;
+					(bonded_pool.try_bond_funds(&who, claimed, BondType::Later)?, claimed)
+				},
+			};
+			delegator.points = delegator.points.saturating_add(points_issued);
+
+			Self::deposit_event(Event::<T>::Bonded {
+				delegator: who.clone(),
+				pool_id: delegator.pool_id,
+				bonded,
+				joined: false,
+			});
+			Self::put_delegator_with_pools(&who, delegator, bonded_pool, reward_pool);
+
+			Ok(())
+		}
+
+		/// A bonded delegator can use this to claim their payout based on the rewards that the pool
+		/// has accumulated since their last claimed payout (OR since joining if this is there first
+		/// time claiming rewards). The payout will be transffered to the delegator's account.
+		///
+		/// The delegator will earn rewards pro rata based on the delegators stake vs the sum of the
+		/// delegators in the pools stake. Rewards do not "expire".
+		#[pallet::weight(T::WeightInfo::claim_payout())]
```

Comment:
```suggestion
#[transactional]
```

let's make everything transactional here?